nvflare/app_opt/tensor_stream/client.py

```diff
@@ -0,0 +1,157 @@
+        """
+        super().__init__()
+        self.format = format
+        self.tasks = tasks if tasks is not None else ["train"]
+        self.enable_request_task_data_tensors = enable_request_task_data_tensors
+        self.wait_for_task_data_tensors_timeout = wait_for_task_data_tensors_timeout
+        self.entry_timeout = entry_timeout
+        self.engine: StreamableEngine = None
+        self.sender: TensorSender = None
+        self.receiver: TensorReceiver = None
+
+    def initialize(self, fl_ctx: FLContext):
+        """Initialize the TensorClientStreamer component.
+        Args:
+            fl_ctx (FLContext): The FLContext for the current operation.
+        """
+        engine: StreamableEngine = fl_ctx.get_engine()
+        if not engine:
+            self.system_panic(f"Engine not found. {self.__class__.__name__} exiting.", fl_ctx)
+            return
+
+        if not isinstance(engine, StreamableEngine):
+            self.system_panic(
+                f"Engine is not a StreamableEngine. {self.__class__.__name__} exiting.",
+                fl_ctx,
+            )
+            return
+
+        self.engine = engine
+        try:
+            self.receiver = TensorReceiver(engine, FLContextKey.TASK_DATA, self.format)
+        except Exception as e:
+            self.system_panic(str(e), fl_ctx)
+            return
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        """Handle events for the TensorSender component.
+
+        Args:
+            event_type (str): The type of event to handle.
+            fl_ctx (FLContext): The FLContext for the current operation.
+        """
+        if event_type == EventType.START_RUN:
+            self.initialize(fl_ctx)
+        elif event_type == EventType.BEFORE_TASK_DATA_FILTER:
+            task_id = fl_ctx.get_prop(FLContextKey.TASK_ID)
+            peer_name = fl_ctx.get_peer_context().get_identity_name()
+            try:
+                if self.enable_request_task_data_tensors:
+                    self.request_task_data_tensors(fl_ctx)
+                    # use longer timeout to wait for tensors from server
+                    # it may take time for server to send all tensors depending on the size and network condition
+                    self.receiver.wait_for_tensors(task_id, peer_name, self.wait_for_task_data_tensors_timeout)
+                else:
+                    # use default short timeout to wait for tensors, data should already received
+                    self.receiver.wait_for_tensors(task_id, peer_name)
+
+                self.receiver.set_ctx_with_tensors(fl_ctx)
+            except Exception as e:
+                self.system_panic(str(e), fl_ctx)
+        elif event_type == EventType.AFTER_TASK_RESULT_FILTER:
+            self.sender = TensorSender(self.engine, FLContextKey.TASK_RESULT, self.format, self.tasks)
+            try:
+                self.send_tensors_to_server(fl_ctx)
```

Comment:
minor stuff, maybe this self.sender = xxx can be put inside the send_tensors_to_server method since we also set to None there